storage: use consistent range descriptor for raft tombstones

tamird opened the pull request:
Fixes #11591.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/11699)
<!-- Reviewable:end -->

tamird committed:
storage: use consistent range descriptor for raft tombstones

Fixes #11591.

petermattis commented:
Definitely needs a test. I've been trying but have so far been unable to reproduce badness here locally.

petermattis commented:
Ok, I got this to fail with a few tweaks to `TestRemovePlaceholderRace`:

```
func TestRemovePlaceholderRace(t *testing.T) {
	defer leaktest.AfterTest(t)()
	sc := storage.TestStoreConfig(nil)
	sc.RaftTickInterval = time.Millisecond
	mtc := &multiTestContext{storeConfig: &sc}
	mtc.Start(t, 3)
	defer mtc.Stop()

	const rangeID = roachpb.RangeID(1)
	mtc.replicateRange(rangeID, 1, 2)

	repl, err := mtc.stores[0].GetReplica(rangeID)
	if err != nil {
		t.Fatal(err)
	}

	ident := mtc.idents[1]
	ctx := repl.AnnotateCtx(context.Background())

	for i := 0; i < 1000; i++ {
		_ = repl.ChangeReplicas(
			ctx,
			roachpb.REMOVE_REPLICA,
			roachpb.ReplicaDescriptor{
				NodeID:  ident.NodeID,
				StoreID: ident.StoreID,
			},
			repl.Desc(),
		)
		_ = repl.ChangeReplicas(
			ctx,
			roachpb.ADD_REPLICA,
			roachpb.ReplicaDescriptor{
				NodeID:  ident.NodeID,
				StoreID: ident.StoreID,
			},
			repl.Desc(),
		)
	}
}
```

---

Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[pkg/storage/replica.go, line 722 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/11699#-KXlvL4T_TRatv5-V6rL:-KXlvL4UoHwxBJuAm6FT:b-nc9hhb) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d194484140a593c1ab82619c65f4e27de014596/pkg/storage/replica.go#L722)):*
> ```Go
> // tombstone. Requires that Replica.raftMu is held.
> func (r *Replica) destroyDataRaftMuLocked(ctx context.Context, desc roachpb.RangeDescriptor) error {
> 	iter := NewReplicaDataIterator(&desc, r.store.Engine(), false /* !replicatedOnly */)
> ```

We should probably leave this using `r.Desc()` as `desc` could have different start/end keys.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11699#-:-KXlvkIpEMe3i0no3V9g:b2pez8u)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Given the tweaks (very small tick interval and ignoring errors from `ChangeReplicas`), this deserves to be a new test.

---

Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11699#-:-KXlvzB1irgneeJra8zG:b-9idig8)*
<!-- Sent from Reviewable.io -->

petermattis commented:
```
func TestReplicaGCRace(t *testing.T) {
	defer leaktest.AfterTest(t)()
	sc := storage.TestStoreConfig(nil)
	sc.RaftTickInterval = time.Millisecond
	mtc := &multiTestContext{storeConfig: &sc}
	mtc.Start(t, 3)
	defer mtc.Stop()

	const rangeID = roachpb.RangeID(1)
	mtc.replicateRange(rangeID, 1, 2)

	repl, err := mtc.stores[0].GetReplica(rangeID)
	if err != nil {
		t.Fatal(err)
	}

	ident := mtc.idents[1]
	ctx := repl.AnnotateCtx(context.Background())

	actions := []roachpb.ReplicaChangeType{roachpb.REMOVE_REPLICA, roachpb.ADD_REPLICA}
	for i := 0; i < 1000; i++ {
		_ = repl.ChangeReplicas(
			ctx,
			actions[i%2],
			roachpb.ReplicaDescriptor{
				NodeID:  ident.NodeID,
				StoreID: ident.StoreID,
			},
			repl.Desc(),
		)
	}
}
```

---

Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11699#-:-KXlwkGJVQ9jvqYB-D2n:b6kr7w3)*
<!-- Sent from Reviewable.io -->

tamird commented:
Updated - I've changed `TestRemovePlaceholderRace` to ignore the irrelevant errors with a TODO since that test is itself flaky.

---

Review status: 0 of 4 files reviewed at latest revision, 1 unresolved discussion.

---

*[pkg/storage/replica.go, line 722 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/11699#-KXlvL4T_TRatv5-V6rL:-KXlySOMHa3BUB9Ersrg:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d194484140a593c1ab82619c65f4e27de014596/pkg/storage/replica.go#L722)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

We should probably leave this using `r.Desc()` as `desc` could have different start/end keys.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11699#-:-KXlyYonp6m8PJuOft7n:bty8wjb)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 4 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[pkg/storage/client_raft_test.go, line 1868 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/11699#-KXm3DCZZq39hPLYzhHM:-KXm3DC_-jOSI-z1VImd:b7b2mli) ([raw file](https://github.com/cockroachdb/cockroach/blob/799816ea8091d787279606595324181e9bb8df5d/pkg/storage/client_raft_test.go#L1868)):*
> ```Go
> 	ctx := repl.AnnotateCtx(context.Background())
> 
> 	for i := 0; i < 100; i++ {
> ```

I couldn't reproduce the failure with this setting. Were you able to reproduce the failure with this test?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11699)*
<!-- Sent from Reviewable.io -->

tamird commented:
With the current test, and the non-test changes backed out:
```
$ make stress PKG=./pkg/storage TESTS=TestRemovePlaceholderRace
go list -tags '' -f 'go test -v  -tags '\'''\'' -ldflags '\'''\'' -i -c {{.ImportPath}} -o {{.Dir}}/stress.test && (cd {{.Dir}} && if [ -f stress.test ]; then stress  ./stress.test -test.run '\''TestRemovePlaceholderRace'\'' -test.timeout 2m ; fi)' ./pkg/storage | /usr/local/bin/bash
github.com/cockroachdb/cockroach/pkg/storage
github.com/cockroachdb/cockroach/pkg/ts
github.com/cockroachdb/cockroach/pkg/server
github.com/cockroachdb/cockroach/pkg/testutils/testcluster
github.com/cockroachdb/cockroach/pkg/storage
github.com/cockroachdb/cockroach/pkg/storage
github.com/cockroachdb/cockroach/pkg/ts
github.com/cockroachdb/cockroach/pkg/server
github.com/cockroachdb/cockroach/pkg/testutils/testcluster
github.com/cockroachdb/cockroach/pkg/storage_test
testmain
0 runs so far, 0 failures, over 5s
1 runs so far, 0 failures, over 10s
1 runs so far, 0 failures, over 15s
2 runs so far, 0 failures, over 20s
4 runs so far, 0 failures, over 25s

I161129 16:51:33.181739 1 rand.go:76  Random seed: 9198666135309503142
I161129 16:51:33.184694 37 storage/engine/rocksdb.go:340  opening in memory rocksdb instance
I161129 16:51:33.186643 37 gossip/gossip.go:244  [n?] initial resolvers: []
W161129 16:51:33.186683 37 gossip/gossip.go:1120  [n?] no resolvers found; use --join to specify a connected node
I161129 16:51:33.186740 37 base/node_id.go:62  NodeID set to 1
I161129 16:51:33.189439 37 storage/store.go:1201  [n1,s1]: failed initial metrics computation: [n1,s1]: system config not yet available
I161129 16:51:33.189555 37 gossip/gossip.go:286  [n1] NodeDescriptor set to node_id:1 address:<network_field:"tcp" address_field:"127.0.0.1:52843" > attrs:<> locality:<>
I161129 16:51:33.190532 54 storage/replica_proposal.go:351  [s1,r1/1:/M{in-ax},@c42022d000] new range lease replica {1 1 1} 1970-01-01 00:00:00 +0000 UTC 9.000124ms following replica {0 0 0} 1970-01-01 00:00:00 +0000 UTC 0s [physicalTime=1970-01-01 00:00:00.000000123 +0000 UTC]
I161129 16:51:33.191049 37 storage/engine/rocksdb.go:340  opening in memory rocksdb instance
I161129 16:51:33.192223 37 gossip/gossip.go:244  [n?] initial resolvers: [127.0.0.1:52843]
W161129 16:51:33.192252 37 gossip/gossip.go:1122  [n?] no incoming or outgoing connections
I161129 16:51:33.192286 37 base/node_id.go:62  NodeID set to 2
I161129 16:51:33.193367 37 storage/store.go:1201  [n2,s2]: failed initial metrics computation: [n2,s2]: system config not yet available
...
I161129 16:51:51.785118 37 storage/replica_command.go:3273  [s1,r1/1:/M{in-ax},@c42022d000] change replicas: read existing descriptor range_id:1 start_key:"" end_key:"\377\377" replicas:<node_id:1 store_id:1 replica_id:1 > replicas:<node_id:3 store_id:3 replica_id:3 > replicas:<node_id:2 store_id:2 replica_id:93 > next_replica_id:94
I161129 16:51:51.841193 19873 storage/replica.go:2192  [s1,r1/1:/M{in-ax},@c42022d000] proposing REMOVE_REPLICA {NodeID:2 StoreID:2 ReplicaID:93}: [{NodeID:1 StoreID:1 ReplicaID:1} {NodeID:3 StoreID:3 ReplicaID:3}]
I161129 16:51:51.861637 37 storage/replica_raftstorage.go:459  [s1,r1/1:/M{in-ax},@c42022d000] generated preemptive snapshot 3e4b6b74 at index 5346
I161129 16:51:51.861918 970 storage/store.go:3015  [s2,r1/?:/M{in-ax},@c4210f6000] added to replica GC queue (peer suggestion)
I161129 16:51:51.863056 970 storage/store.go:3015  [s2,r1/?:/M{in-ax},@c4210f6000] added to replica GC queue (peer suggestion)
I161129 16:51:51.863386 155 storage/store.go:1991  [replicaGC,s2,r1/?:/M{in-ax},@c4210f6000] removing replica
I161129 16:51:51.919465 55 storage/replica.go:2239  [s1,r1/1:/M{in-ax},@c42022d000] not quiescing: 2 pending commands
E161129 16:51:51.933112 121 vendor/github.com/coreos/etcd/raft/log.go:191  [s2,r1/93:{-},@c42102a000] tocommit(5336) is out of range [lastIndex(0)]. Was the raft log corrupted, truncated, or lost?
panic: tocommit(5336) is out of range [lastIndex(0)]. Was the raft log corrupted, truncated, or lost? [recovered]
	panic: tocommit(5336) is out of range [lastIndex(0)]. Was the raft log corrupted, truncated, or lost?

...

ERROR: exit status 2

6 runs completed, 1 failures, over 29s
FAIL
make: *** [stress] Error 1
```

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 4 files reviewed at latest revision, 3 unresolved discussions.

---

*[pkg/storage/client_raft_test.go, line 1882 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/11699#-KXm4PaFJpxNOGgRThdx:-KXm4PaFJpxNOGgRThdy:b-rgth3m) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb3894c4f31ac10d0593c957a838fef67c76082f/pkg/storage/client_raft_test.go#L1882)):*
> ```Go
> 					break
> 				} else if testutils.IsError(err, "cannot add placeholder, have an existing placeholder") {
> 					break // TODO(tamird): why is this clause necessary?
> ```

After the `AmbiguousResultError` would a subsequent `ADD_REPLICA` cause this error?

---

*[pkg/storage/replica.go, line 721 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/11699#-KXm4iGcLCdcGYNCtOip:-KXm4iGdMci9ZxvutH99:b-9zjrr3) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb3894c4f31ac10d0593c957a838fef67c76082f/pkg/storage/replica.go#L721)):*
> ```Go
> // destroyData deletes all data associated with a replica, leaving a
> // tombstone. Requires that Replica.raftMu is held.
> func (r *Replica) destroyDataRaftMuLocked(ctx context.Context, desc roachpb.RangeDescriptor) error {
> ```

`s/desc/consistentDesc/g`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11699#-:-KXm4x0m5dTQsSp9uWjl:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: 0 of 4 files reviewed at latest revision, 3 unresolved discussions.

---

*[pkg/storage/client_raft_test.go, line 1868 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/11699#-KXm3DCZZq39hPLYzhHM:-KXmAC9AxO5rlHWJAP5l:bb77ufg) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb3894c4f31ac10d0593c957a838fef67c76082f/pkg/storage/client_raft_test.go#L1868)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I couldn't reproduce the failure with this setting. Were you able to reproduce the failure with this test?
</blockquote></details>

Yes.

---

*[pkg/storage/client_raft_test.go, line 1882 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/11699#-KXm4PaFJpxNOGgRThdx:-KXmABJiBLvMCT29utCJ:bss4a2b) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb3894c4f31ac10d0593c957a838fef67c76082f/pkg/storage/client_raft_test.go#L1882)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

After the `AmbiguousResultError` would a subsequent `ADD_REPLICA` cause this error?
</blockquote></details>

Removed.

---

*[pkg/storage/replica.go, line 721 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/11699#-KXm4iGcLCdcGYNCtOip:-KXmADISUCYZf7f3969x:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb3894c4f31ac10d0593c957a838fef67c76082f/pkg/storage/replica.go#L721)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

`s/desc/consistentDesc/g`?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11699)*
<!-- Sent from Reviewable.io -->

tamird committed:
build: correct crlfmt instructions

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 5 files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11699#-:-KXmF3ChaNDsC25G8194:bnfp4nl)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Let's get this merged. I'm 99% sure the reproposal loop is unrelated to this PR, though it is really nice we have a reproducible test case for that now.

tamird commented:
Yeah, I'm trying. Rebuild in progress.

petermattis commented:
Adding `--vmodule=raft=4` shows that conflicting elections are happening in that test. Setting `sc.RaftElectionTimeoutTicks = 1000000` removes the flakiness. Unfortunately, if you set `RaftElectionTimeoutTicks` high the original error does not reproduce. At least, it doesn't for me. Perhaps you'll have more luck.

tamird commented:
Grrr. I'll take a look after dinner.

petermattis commented:
Review status: 0 of 5 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[pkg/storage/client_raft_test.go, line 1855 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/11699#-KXmTsEAVlVFA6md9r9m:-KXmTsEAVlVFA6md9r9n:bsqpqbq) ([raw file](https://github.com/cockroachdb/cockroach/blob/f91562db184310c40a0ba036cb61db07432c0858/pkg/storage/client_raft_test.go#L1855)):*
> ```Go
> 
> 	const rangeID = roachpb.RangeID(1)
> 	mtc.replicateRange(rangeID, 1, 2)
> ```

I think this is the source of the trouble. Instead of having 3 replicas you only have 2. Revert this line and then reverse the order of the add and remove below. Then repeat: "I will not make spurious cleanups when working on a high priority PR".

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11699)*
<!-- Sent from Reviewable.io -->

tamird commented:
D: I had to make that change to work around the random placeholder errors

On Nov 29, 2016 6:49 PM, "Peter Mattis" <notifications@github.com> wrote:

> Review status: 0 of 5 files reviewed at latest revision, 1 unresolved
> discussion, some commit checks failed.
> ------------------------------
>
> *pkg/storage/client_raft_test.go, line 1855 at r4
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/11699#-KXmTsEAVlVFA6md9r9m:-KXmTsEAVlVFA6md9r9n:bsqpqbq>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/f91562db184310c40a0ba036cb61db07432c0858/pkg/storage/client_raft_test.go#L1855>):*
>
> 	const rangeID = roachpb.RangeID(1)
> 	mtc.replicateRange(rangeID, 1, 2)
>
> I think this is the source of the trouble. Instead of having 3 replicas
> you only have 2. Revert this line and then reverse the order of the add and
> remove below. Then repeat: "I will not make spurious cleanups when working
> on a high priority PR".
> ------------------------------
>
> *Comments from Reviewable
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/11699>*
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/11699#issuecomment-263737937>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABdsPAV2o6YORLxkJfV0ZHhCyDZ3by5sks5rDLn4gaJpZM4K_eYW>
> .
>

petermattis commented:
> D: I had to make that change to work around the random placeholder errors

Perhaps it is worth forking the test. I don't see value in trying to be overly strict in which errors we expect. In other words, lets leave `TestRemovePlaceholderRace` as is and add `TestReplicaGCRace` as I created above. And then add a comment about why the errors from `ChangeReplicas` are being intentionally ignored.

petermattis commented:
> Perhaps it is worth forking the test.

Sigh. I just tried stressing `TestReplicaGCRace` and it is flaky as well. I'll revise my suggestion: perhaps we should punt on creating a reproducible test until after this PR is merged.

---

Review status: 0 of 5 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11699#-:-KXn8mWVjzYTgWacs8Vp:ba2r6uo)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

---

Reviewed 1 of 5 files at r3, 4 of 4 files at r4.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/storage/client_raft_test.go, line 1855 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/11699#-KXmTsEAVlVFA6md9r9m:-KXnhewzWRTx3xoceN4E:b-qt2xjj) ([raw file](https://github.com/cockroachdb/cockroach/blob/f91562db184310c40a0ba036cb61db07432c0858/pkg/storage/client_raft_test.go#L1855)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I think this is the source of the trouble. Instead of having 3 replicas you only have 2. Revert this line and then reverse the order of the add and remove below. Then repeat: "I will not make spurious cleanups when working on a high priority PR".
</blockquote></details>

Yeah, removing this line changes the test since the node that is added/removed will have to participate in the quorum. But I think for this test the two-replica form is what we want (otherwise we lose synchronization and we could just zip through our 100 iterations while skipping "expected" errors). 

---

*[pkg/storage/client_raft_test.go, line 1879 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/11699#-KXnifWny-uMXXJrH3dm:-KXnifWny-uMXXJrH3dn:b557fzd) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad59486257a4b2e59dc2265da6b231450df5eeb8/pkg/storage/client_raft_test.go#L1879)):*
> ```Go
> 				// TODO(tamird): why is this clause necessary?
> 				if testutils.IsError(err, "cannot add placeholder, have an existing placeholder") {
> 					break // skip the removal, we've hit some kind of race.
> ```

This error would occur on additions, not removals. I think it would be expected in the three-replica case and unexpected in the two-replica case. Is this still needed with the replicateRange line removed? If we do need to break here I think we need to do `i--` to ensure that the test doesn't trivially pass without actually doing anything. 

---

*[pkg/storage/client_test.go, line 906 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/11699#-KXnfeGwgi0KDeg1HLXA:-KXnfeGwgi0KDeg1HLXB:b-4rp7em) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad59486257a4b2e59dc2265da6b231450df5eeb8/pkg/storage/client_test.go#L906)):*
> ```Go
> 		}
> 
> 		ctx := repl.AnnotateCtx(context.Background())
> ```

Why is this context.Background and the previous one context.TODO? Seems to me that they should both be the same. 

---

*[pkg/storage/replica.go, line 724 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/11699#-KXnegWmp15U6h3BBr8g:-KXnegWmp15U6h3BBr8h:bouv8e5) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad59486257a4b2e59dc2265da6b231450df5eeb8/pkg/storage/replica.go#L724)):*
> ```Go
> 	ctx context.Context, consistentDesc roachpb.RangeDescriptor,
> ) error {
> 	iter := NewReplicaDataIterator(r.Desc(), r.store.Engine(), false /* !replicatedOnly */)
> ```

Comment that `r.Desc()` matches the data we have on disk so we use it instead of consistentDesc for the ReplicaDataIterator.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11699#-:-KXnk2HwbwfiVK-xqf09:b-uwkowd)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
The proposal loop (which is different from what we're seeing in #11667) appears to be triggered by the use of two replicas instead of three, aggravated by the reduction in RaftTickInterval. Here's what I found when I reproduced the problem with `--vmodule=raft=4`:

Under stress, the reduced RaftTickInterval causes the leader (store 1) to sometimes step down due to the raft CheckQuorum feature (this part will be changed when we switch from CheckQuorum to PreVote, but the fundamental problem will remain).

```
W161130 16:23:56.738025 64 vendor/github.com/coreos/etcd/raft/raft.go:794  [s1,r1/1:/M{in-ax},@c4201fb000] 1 stepped down to follower since quorum is not active
I161130 16:23:56.738046 64 vendor/github.com/coreos/etcd/raft/raft.go:568  [s1,r1/1:/M{in-ax},@c4201fb000] 1 became follower at term 142
I161130 16:23:56.738100 64 vendor/github.com/coreos/etcd/raft/raft.go:1029  [s1,r1/1:/M{in-ax},@c4201fb000] 1 no leader at term 142; dropping proposal
```

This leads to elections, and store 2 wins one of them:

```
I161130 16:23:56.741197 113 vendor/github.com/coreos/etcd/raft/raft.go:622  [s2,r1/39:/M{in-ax},@c420ab9400] 27 became leader at term 143
```

Store 2 writes an empty entry to its log, but before it can broadcast the MsgApp to the follower on store 1, store 1 times out and starts a new election:

```
I161130 16:23:56.754688 64 vendor/github.com/coreos/etcd/raft/raft.go:750  [s1,r1/1:/M{in-ax},@c4201fb000] 1 is starting a new election at term 143
I161130 16:23:56.754719 64 vendor/github.com/coreos/etcd/raft/raft.go:581  [s1,r1/1:/M{in-ax},@c4201fb000] 1 became candidate at term 144
```

Store 1 can't win this election because its log is behind store 2's, but the fact that it's trying is enough to make store 2 step down (again, PreVote will change this part but problems will still be possible with the right timing):

```
I161130 16:23:56.786587 113 vendor/github.com/coreos/etcd/raft/raft.go:710  [s2,r1/39:/M{in-ax},@c420ab9400] 27 [term: 144] received a MsgVote message with higher term from 1 [term: 145]
I161130 16:23:56.786607 113 vendor/github.com/coreos/etcd/raft/raft.go:568  [s2,r1/39:/M{in-ax},@c420ab9400] 27 became follower at term 145
I161130 16:23:56.786648 113 vendor/github.com/coreos/etcd/raft/raft.go:774  [s2,r1/39:/M{in-ax},@c420ab9400] 27 [logterm: 143, index: 1388, vote: 0] rejected MsgVote from 1 [logterm: 143, index: 1387] at term 145
I161130 16:23:56.787268 59 vendor/github.com/coreos/etcd/raft/raft.go:667  [s1,r1/1:/M{in-ax},@c4201fb000] 1 received MsgVoteResp rejection from 27 at term 145
I161130 16:23:56.787313 59 vendor/github.com/coreos/etcd/raft/raft.go:1008  [s1,r1/1:/M{in-ax},@c4201fb000] 1 [quorum:2] has received 1 MsgVoteResp votes and 1 vote rejections
```

Now we have the real problem: store 2 is the only replica that can win an election, but it won't campaign (store 1 will keep campaigning and failing until time runs out). Store 2 is clearly getting awoken by store 1's MsgVotes, and it shouldn't be quiescing because there is no leader. 

My best guess as to what's happening is that when store 1 stepped down, there were replica changes in the tail of the log. These entries had been appended to store 2's log but store 1 didn't get the ack in time so it never marked the entries as committed. When store 2 won the election, it found that the entries were in fact committed, so it removed itself and considered itself to no longer be a valid candidate. In its last action as leader, it made a single attempt at broadcasting the final log entries, which was rejected because store 1 had already incremented its term:

```
I161130 16:23:56.756950 116 vendor/github.com/coreos/etcd/raft/raft.go:710  [s2,r1/39:/M{in-ax},@c420ab9400] 27 [term: 143] received a MsgAppResp message with higher term from 1 [term: 144]
```

This problem is inherent in the process of downreplicating from two nodes to one: there is no way to guarantee that the remaining node ends up being fully informed. That's why this test (like all tests that downreplicate) use three replicas, although I had forgotten this. 

We should restore this test to use three replicas and the default RaftTickInterval, and then figure out how to reliably test this change after it's merged.

tamird commented:
Done.

---

Review status: 1 of 3 files reviewed at latest revision, 4 unresolved discussions.

---

*[pkg/storage/replica.go, line 724 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/11699#-KXnegWmp15U6h3BBr8g:-KXpMQHkhqjE9IrQAeLu:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad59486257a4b2e59dc2265da6b231450df5eeb8/pkg/storage/replica.go#L724)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Comment that `r.Desc()` matches the data we have on disk so we use it instead of consistentDesc for the ReplicaDataIterator.
</blockquote></details>

Done.

---

*[pkg/storage/client_raft_test.go, line 1855 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/11699#-KXmTsEAVlVFA6md9r9m:-KXpLMufFP1TM-nbesqR:bp3aow6) ([raw file](https://github.com/cockroachdb/cockroach/blob/f91562db184310c40a0ba036cb61db07432c0858/pkg/storage/client_raft_test.go#L1855)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Yeah, removing this line changes the test since the node that is added/removed will have to participate in the quorum. But I think for this test the two-replica form is what we want (otherwise we lose synchronization and we could just zip through our 100 iterations while skipping "expected" errors). 
</blockquote></details>

Reverted.

---

*[pkg/storage/client_raft_test.go, line 1879 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/11699#-KXnifWny-uMXXJrH3dm:-KXpLLyiQbfvwkbspH2X:bp3aow6) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad59486257a4b2e59dc2265da6b231450df5eeb8/pkg/storage/client_raft_test.go#L1879)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

This error would occur on additions, not removals. I think it would be expected in the three-replica case and unexpected in the two-replica case. Is this still needed with the replicateRange line removed? If we do need to break here I think we need to do `i--` to ensure that the test doesn't trivially pass without actually doing anything. 
</blockquote></details>

Reverted.

---

*[pkg/storage/client_test.go, line 906 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/11699#-KXnfeGwgi0KDeg1HLXA:-KXpLkQtWGDA-yKBEY1I:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad59486257a4b2e59dc2265da6b231450df5eeb8/pkg/storage/client_test.go#L906)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Why is this context.Background and the previous one context.TODO? Seems to me that they should both be the same. 
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11699#-:-KXpLJ7oQHPFsgyvXogb:b-896fix)*
<!-- Sent from Reviewable.io -->
